File: papers/aplas/benchmarks/maxLengtho/original.pl
maxLengtho(X, M, L) :- maxo(X, M), lengtho(X, L).
maxo(X, M) :- maxo1(X, o, M).
maxo1(nil, N, N).
maxo1(cons(H, T), N, M) :- leo(H, N, true), maxo1(T, N, M).
maxo1(cons(H, T), N, M) :- gto(H, N, true), maxo1(T, H, M).
leo(o, Y, true).
leo(s(Zz), o, false).
leo(s(X1), s(Y1), B) :- leo(X1, Y1, B).
gto(s(Zz), o, true).
gto(o, Y, false).
gto(s(X1), s(Y1), B) :- gto(X1, Y1, B).
lengtho(nil, o).
lengtho(cons(H, T), s(Z)) :- lengtho(T, Z).

File: papers/aplas/benchmarks/maxLengtho/original.ml
open GT
open OCanren
open OCanren.Std
open Helper

let topLevel x0 x1 x2 =
let rec maxLengtho x m l = ((maxo x m) &&& (lengtho x l))
and maxo x m = (maxo1 x Std.Nat.zero m)
and maxo1 x n m = (fresh (z t h) (((x === (Std.List.nil ())) &&& (m === n)) ||| ((((x === (Std.(%) (h) (t))) &&& ((leo h n !!true) &&& (maxo1 t n m)))) ||| ((((x === (Std.(%) (h) (t))) &&& ((gto h n !!true) &&& (maxo1 t h m))))))))
and leo x y b = fresh (zz x' y') ((((x === Std.Nat.zero) &&& (b === !!true)) ||| (((((x === (Std.Nat.succ (zz))) &&& ((y === Std.Nat.zero) &&& (b === !!false))))) ||| ((((x === (Std.Nat.succ (x'))) &&& ((y === (Std.Nat.succ (y'))) &&& (leo x' y' b))))))))
and gto x y b = fresh (zz x' y')  (((((x === (Std.Nat.succ (zz))) &&& ((y === Std.Nat.zero) &&& (b === !!true))))) ||| (((x === Std.Nat.zero) &&& (b === !!false)) ||| ((((x === (Std.Nat.succ (x'))) &&& ((y === (Std.Nat.succ (y'))) &&& (gto x' y' b)))))))
and lengtho x l = fresh (z t h) ((((x === (Std.List.nil ())) &&& (l === Std.Nat.zero)) ||| ( (((x === (Std.(%) (h) (t))) &&& ((l === (Std.Nat.succ (z))) &&& (lengtho t z))))))) in       ( ((maxLengtho x0 x1 x2 )))
